### ISSUE TITLE
Managed to crash netdisco weboutput

### ISSUE THREAD

**Issue by zoeloe** [netdisco/netdisco#610]
I'm not proud of it, but I think a certain search query crashes netdisco weboutput.

## Expected Behavior
Resist crashing when using malformed searches

## Current Behavior
Crash, generated weboutput wrong
crashdump in gist:
https://gist.github.com/zoeloe/086bf4b10ceace6f5f1c6c389a45195e

## Possible Solution
Filter malformed queries?

## Steps to Reproduce (for bugs)
1. http://<host.domain.part1.td>/search?q=2C101%2FA%2F5+2C101+++2C101%2FA%2F6+2C101&firstsearch=on

## Context
pasted a copy buffer by accident in the main search field on home page. 

## Your Environment
Ubuntu 16.04.05
* Netdisco version used: 2.42.10
* SNMP::Info version used: 3.68

## Device information
not relevant.

**Comment by inphobia** [comment]
it's indeed easy to reproduce. a fix however seems less evident since the search bar tries to play it smart to figure out if you're looking for vlans, nodes, devices or ports. so fixing the mac detection seems to have unintended side effects.

general input sanitation seems to be a good path to take, that or better error handling. would love some input from @netdisco/netdisco-admins / @netdisco/netdisco-developers so see what their take is.

(side note, am slowly working on a frontend rewrite, this could be a target for that too)

**Comment by inphobia** [comment]
trouwens, blij om te zien dat de vub bij onze gebruikers zit :)

**Comment by zoeloe** [comment]
@inphobia 
Nice to know that there is a dutch speaking (Like me !) and a major dev/contributor on the netdisco project.
The only mini contrib from me was in #239 for SNMP::Info.
And some bug reports which I bumped into.

You can leave your personal coordinates for correspondence in the dutch language if you like, for me.
(replied in English for general readability :) )

Yes, I work at the VUB university and promoted the use of netdisco from the early existence on, because of the (IMHO) layered (SNMP::Info abstraction layer) philosophy as starters.

Keep up the great work. I certainly appreciate.

Regards, Thanks for your contributions.

**Comment by JeroenvIS** [comment]
@zoeloe inphobia isn't the only Dutch speaking dev on the project ;-) (though I'm not contributing much these days...)
But indeed, let's try to keep the tickets and other general correspondence in English.

**Comment by JeroenvIS** [comment]
> general input sanitation seems to be a good path to take, that or better error handling. would love some input from @netdisco/netdisco-admins / @netdisco/netdisco-developers so see what their take is.

IMHO we should always try our best to do both: sanitize input where possible and gracefully handle all errors that may arise.
In this specific case: since we know what syntax is acceptable to Postgres for macaddr type, we should make sure that we submit a valid MAC in valid notation to the database.

**Comment by zoeloe** [comment]
Mhhh,.... IMHO input sanitizing  will be very difficult, as e.g. in our installation we default to enable "Partial Match" "Uplinks" and "Ethernet Only" search in order to actually search anything you throw at the search field, throughout the database.

sidebar_defaults:
  search_port:
    partial:      { default: checked }
    uplink:       { default: checked }
    ethernet:     { default: checked }

(We have a lot of ports labelled with some info, which is handy to have in the netdisco database, available to search.)

But that's our usage case... Any other opinions ?

**Comment by inphobia** [comment]
> In this specific case: since we know what syntax is acceptable to Postgres for macaddr type, we should make sure that we submit a valid MAC in valid notation to the database.

the thing is, we're trying to be smart when using the default search field and guessing if we want to look for vlans/nodes/devices/ports:
https://github.com/netdisco/netdisco/blob/6f8dfabc87f14b65bfd8f87aff85b07f8f8cea11/lib/App/Netdisco/Web/Search.pm#L34-L42

making sure the mac is valid for postgres isn't the biggest challenge i think (netaddr::mac even has a function for this: https://metacpan.org/pod/NetAddr::MAC#as_pgsql), but guessing what the user is trying to find is.

validating for all 4 search options seems like a challenge to me.

**Comment by inphobia** [comment]
> Mhhh,.... IMHO input sanitizing will be very difficult, as e.g. in our installation we default to enable "Partial Match" "Uplinks" and "Ethernet Only" search in order to actually search anything you throw at the search field, throughout the database.

are you mostly using port search (as in: you just use the search bar on top of the netdisco page but most of the time expect a result on the "port" page?) 

> But that's our usage case... Any other opinions ?

i don't think we currently have a way to change search order anywhere. custom reports could be an option, but are less userfriendly: https://github.com/netdisco/netdisco/wiki/Configuration#reports

improving https://metacpan.org/pod/NetAddr::MAC#ERROR-HANDLING error handling is also a possible option.

perhaps someone has a better idea to make fuzzy searches & input validation work together?

**Comment by zoeloe** [comment]
We mostly search for IP addresses and in the results, there are clickable MAC adresses (which DO have a correct format)..  The search for others patterns was an idea in testing and worked as expected till the accidental bug discovery.
If the options partial: { default: checked }
uplink: { default: checked }
ethernet: { default: checked }
are enabled, IMHO the behaviour of the top search bar and the "central" home screen "Find anything" are the same. They show the results on the "Port Serach Options" page.

Not using custom reports at all.

So yes, maybe an error capture would be advised for the MAC address detection. ?
(Which I guess triggered the error ?)
Or simply avoid crashing out when an  error is thrown  ?

**Comment by ollyg** [comment]
Thanks! Fixed in
[master ceeccf89] #610 better MAC addr input sanity checking

We already did have some input sanity checks but in part delegated this to NetAddr::MAC. Arguably NetAddr::MAC should not create such weird addresses as below, but since we recently added Regexp::Common as a dependency, we can use that instead. Also as I am COMAINT on NetAddr::MAC it's my problem either way ;-).

`2c101:0a:05:2c101:2c101:0a:06:2c101`
